### PR TITLE
coll: TSP allgather ring

### PR DESCRIPTION
**Pull request by tarudoodi**

### CODE REVIEW

**Comment by tarudoodi** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

File: src/mpi/coll/iallgather/iallgather_tsp_ring_algos.h [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+    MPIR_Type_get_true_extent_impl(sendtype, &sendtype_lb, &sendtype_true_extent);
+    sendtype_extent = MPL_MAX(sendtype_extent, sendtype_true_extent);
+
+    MPIR_Datatype_get_extent_macro(recvtype, recvtype_extent);
+    MPIR_Type_get_true_extent_impl(recvtype, &recvtype_lb, &recvtype_true_extent);
+    recvtype_extent = MPL_MAX(recvtype_extent, recvtype_true_extent);
+
+    /* Allocate space for temporary buffers */
+    buf1 = MPIR_TSP_sched_malloc(recvcount * recvtype_extent, sched);
+    buf2 = MPIR_TSP_sched_malloc(recvcount * recvtype_extent, sched);
```
Comment (hzhou) [inline src/mpi/coll/iallgather/iallgather_tsp_ring_algos.h]: @tarudoodi I don't quite understand why we can't directly use the `recvbuf` directly here. Could you explain?

File: src/mpi/coll/iallgather/iallgather_tsp_ring_algos.h [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIR_TSP_IALLGATHER_SCHED_INTRA_RING);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIR_TSP_IALLGATHER_SCHED_INTRA_RING);
+
+    /* Find out the buffer which has the send data and point data_buf to it */
+    if (is_inplace) {
+        sendcount = recvcount;
+        sendtype = recvtype;
+        data_buf = (char *) recvbuf;
+    } else
+        data_buf = (char *) sendbuf;
```
Comment (hzhou) [inline src/mpi/coll/iallgather/iallgather_tsp_ring_algos.h]: `data_buf` is only used during `localcopy` and we have to switch on `is_inplace` there anyway. Please move it to below.

**Comment by hzhou** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Pull request merged**